### PR TITLE
fix(atomic): use event to resolve parent result

### PR DESCRIPTION
**Pull request by ThibodeauJF**
The previous method updating the component directly caused weird issues, using the events (same method as for bindings) prevent those issues.
https://coveord.atlassian.net/browse/KIT-604

### CODE REVIEW

File: packages/atomic/package.json [hunk 1]
```diff
@@ -61,6 +62,7 @@
     "jest": "^26.4.2",
     "jest-cli": "^26.4.2",
     "postcss-import": "^14.0.0",
+    "puppeteer": "^9.1.1",
```
Comment (ThibodeauJF) [inline packages/atomic/package.json:65]: The removal of pupeteer, although useless, caused random errors to appear in the dev build 
Screen Shot 2021-05-05 at 2 45 37 PM

File: packages/atomic/src/components/atomic-result/atomic-result.tsx [hunk 2]
```diff
@@ -28,6 +28,13 @@ export class AtomicResult {
    */
   @Prop() content!: string;
 
+  @Listen('atomic/resolveResult')
```
Comment (ThibodeauJF) [inline packages/atomic/src/components/atomic-result/atomic-result.tsx:31]: We have to use primitive always for decorator values, can't use a variable

File: packages/atomic/package.json [hunk 3]
```diff
@@ -29,8 +29,8 @@
     "npm:publish": "npm publish --access public"
   },
   "dependencies": {
-    "@coveo/bueno": "^0.13.0-alpha.5",
-    "@coveo/headless": "^0.13.0-alpha.5",
+    "@coveo/bueno": "^0.12.0-alpha.0",
```
Comment (ThibodeauJF) [inline packages/atomic/package.json:32]: the alpha.5 didn't exist for either packages.

Comment (ThibodeauJF) [reply packages/atomic/package.json:32]: Deployment wasn't finished when I was installing

File: packages/atomic/src/components/result-template-components/result-template-decorators.ts [hunk 4]
```diff
@@ -1,25 +1,35 @@
...
+    );
+  }
+}
 
 export function ResultContext() {
   return (component: ComponentInterface, resultVariable: string) => {
-    const {componentWillLoad} = component;
-    component.componentWillLoad = function () {
+    const {connectedCallback} = component;
+    component.connectedCallback = function () {
```
Comment (olamothe) [inline packages/atomic/src/components/result-template-components/result-template-decorators.ts:16]: Just a sanity check, but is there any chance of a race condition where we trigger the event before the parent would be ready to `Listen` to it ?

I'm assuming not, but just making sure.

File: packages/atomic/src/components/result-template-components/result-template-decorators.ts [hunk 5]
```diff
@@ -1,25 +1,35 @@
+import {Result} from '@coveo/headless';
 import {ComponentInterface, getElement} from '@stencil/core';
+import {buildCustomEvent} from '../../utils/event-utils';
+
+export class MissingResultParentError extends Error {
```
Comment (olamothe) [inline packages/atomic/src/components/result-template-components/result-template-decorators.ts:5]: What is the benefit of making this a class versus before ?

Just for code clarity sake ?

**Review by olamothe — APPROVED** [review]
What is the "weird issue" you are talking about exactly ? 
Can you describe a bit more/go into more details ?

And since it was originally opened for the printable URI component, is there any test that we could add to verify the behaviour ?

File: packages/atomic/src/components/result-template-components/result-template-decorators.ts (hunk 5 above)
Comment (ThibodeauJF) [reply packages/atomic/src/components/result-template-components/result-template-decorators.ts:5]: Yes

File: packages/atomic/src/components/result-template-components/result-template-decorators.ts (hunk 4 above)
Comment (ThibodeauJF) [reply packages/atomic/src/components/result-template-components/result-template-decorators.ts:16]: The parent is the atomic result which renders the content., from a string, previous lifecycle methods will have been launched.
It's different than the atomic-search-interface & other components relationship, which are slotted element.

**Review by samisayegh — APPROVED** [review]

**Comment by ThibodeauJF** [comment]
@olamothe I believe the issue stems from using component[resultVariable] instead of this[resultVariable], the result variable was being overriden when other components were using the decorator, so when we expanded the printable-uri, it would then render the latest assigned result instead of it’s originally assigned result. The behaviour would change depending on which lifecycle method was used, using connectedCallback would make all the components render the same result..

I decided to also switch to the event system since the behaviour is similar to the bindings, and event is a recommended way of bubbling up to parents, instead of manually resolving the atomic-result parent, which is less scalable (what is we have other result components in the future?)

The cypress tests for Printable URI are written but not implemented yet, I imagine we could have caught it with one of those tests.

**Pull request merged**